src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+
+Summary:
+
+ * "Structured" concurrency makes multi-threading much *easier*.
+    * Threads can access data from the containing scope!
+ * "Fearless" concurrency *prevents data races* between threads.
+ * *Structured fearless concurrency combines the best of both worlds.*
+ * Vale's `parallel` keyword makes it happen!
+
+
+# We Want Structured Concurrency!
+
+
+Structured concurrency is a lot of fun.
+
+
+With structured concurrency, you can launch multiple threads at once, to run your calculations in parallel, for a big speedup. And often, it can be done with a tiny adjustment.
+
+
+Imagine we had this simple program that calculated x^3 for some values of x:
+
+```c
+#include <math.h>
+#include <stdio.h>
+
+void main() {
+  int exponent = 3;
+  int results[5];
+
+  // Calculate some exponents
+  for (int i = 0; i < 5; i++) {
+   results[i] = pow(i, exponent);
+  }
+
+  // Print out the results
+  for (int i = 0; i < 5; i++) {
+   printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+}
+```stdout
+0^3 is 0!
+1^3 is 1!
+2^3 is 8!
+3^3 is 27!
+4^3 is 64!
+```
+
+
+
+If `pow` was much more expensive, we might want to run those calculations in parallel.
+
+It's very tedious to run it in C. We'd have to wrap our `results[i] = pow(i, exponent);` line in an entire new function, and use `pthread_create`, like so:
+
+
+```c
+#include <math.h>
+#include <stdio.h>
+#include <pthread.h>
+
+typedef struct {
+  int exponent;
+  int i;
+  int* results;
+} MyThreadArgs;
+
+void *my_thread_main(void* args_raw) {
+  MyThreadArgs* args = (MyThreadArgs*)args_raw;
+  int exponent = args->exponent;
+  int i = args->i;
+  int* results = args->results;
+
+  results[i] = pow(i, exponent); // some expensive calculation
+
+  free(args_raw);
+  return NULL;
+}
+
+int main() {
+  int exponent = 3;
+
+  int results[5];
+
+  // Don't run each iteration one after the other...
+  // run them in parallel, rather than serially.
+  pthread_t threads[5];
+  for (int i = 0; i < 5; i++) {
+    MyThreadArgs* args = (MyThreadArgs*)malloc(sizeof(MyThreadArgs));
+    args->exponent = exponent;
+    args->i = i;
+    args->results = results;
+    pthread_create(&threads[i], NULL, my_thread_main, args);
+  }
+  // Join the threads
+  for (int i = 0; i < 5; i++) {
+    void* returnval = NULL;
+    pthread_join(threads[i], &returnval);
+  }
+
+  for (int i = 0; i < 5; i++) {
+    printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+  return 0;
+}
+```
+
+
+We needed a lot of code to just run `results[i] = pow(i, exponent);` in parallel!
+
+With structured concurrency, we can do that with *just one line*. Let's add a `#pragma omp parallel for` to our original program: [# To run this, you might need to install OpenMP first. For my Mac, I needed to run `brew install libomp` and compile with `clang main.c -fopenmp -L/usr/local/Cellar/libomp/13.0.0/lib`.]
```

Comment:
Just link to the site, don't explain how to install dependancies in the post

Reply:
done